desc/protoparse/source_positions.go

```diff
@@ -0,0 +1,105 @@
+package protoparse
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// ErrorWithPos is an error about a proto source file that includes information
+// about the location in the file that caused the error.
+type ErrorWithPos interface {
+	error
+	GetPosition() SourcePos
+}
+
+// ErrorWithSourcePos is an error about a proto source file that includes
+// information about the location in the file that caused the error. Errors
+// that include source location information *might* be of this type. However,
+// calling code that is trying to examine errors with location info should
+// instead look for instances of the ErrorWithPos interface, which will find
+// other kinds of errors. This type is only exported for backwards
+// compatibility.
+type ErrorWithSourcePos struct {
+	Underlying error
+	Pos        *SourcePos
+}
+
+// Error implements the error interface
+func (e ErrorWithSourcePos) Error() string {
+	if e.Pos.Line <= 0 || e.Pos.Col <= 0 {
+		return fmt.Sprintf("%s: %v", e.Pos.Filename, e.Underlying)
+	}
+	return fmt.Sprintf("%s:%d:%d: %v", e.Pos.Filename, e.Pos.Line, e.Pos.Col, e.Underlying)
+}
+
+// GetPosition implements the ErrorWithPos interface, supplying a location in
+// proto source that caused the error.
+func (e ErrorWithSourcePos) GetPosition() SourcePos {
+	return *e.Pos
```

Comment:
This seems dangerous to me, why not just have `GetPosition` return `*SourcePos` and then you can check for nil? Or make `Pos a `SourcePos`?